### DISCUSSION TITLE
[Feature request] Use a consistent API

### DISCUSSION THREAD

**Discussion by matthew-dean** [primefaces/community#1612] [PrimeVue]
## Form inputs

One thing that was a surprise to me was that some inputs, like `<Button>` and `<InputText>` have a `size` property, yet form inputs like `<Dropdown>` do not. This is counter-intuitive because you cannot consistently size inputs in a row by applying a size. Ideally, all inputs have a set of shared properties that can guarantee consistent styling and sizing.

## Panels and Menus

Here, the API is really all over the place.
- An OverlayPanel's visibility is controlled via a `ref`, in which a method (like `.toggle()` is called)
- A Dialog's visiblity is controlled via its `visible` property.
- A Menu's visibility is controlled via a `ref`
- A ConfirmPopup's visiblity is controlled via... a useConfirm with an empty `<ConfirmPopup>` (what is even going on here)
- A DynamicDialog is similar to ConfirmPopup, but it's own "container" element
- A Toast is similar to ConfirmPopup and DynamicDialog, with its own "container"
- A Tooltip's visibility is controlled by declarative directives

### Proposal for panels and menus
1. Visibility should be both declarative (via a prop) and imperative (via methods)
2. I really don't understand the need for distinct, app-level containers. A single top-level `<OverlayManager>` should suffice
3. Similar to Vuetify, one should not have to "externally wire" activators for popups, e.g.

An ideal API
```vue
<OverlayPanel>
  <template #activator="{ toggle }">
    <Button @click="toggle">Activate Menu</Button>
  </template>

   <div>
     Other content
   </diiv>
</OverlayPanel>
```
The above ☝️  would be fairly easy for a user to do if PrimeVue components could be abstracted, [which they cannot be](https://github.com/orgs/primefaces/discussions/860#discussioncomment-8905929).

**Comment by wijzijnweb** [comment]
A trigger / activator slot for dialogs would be also very nice.
This removes the need to have a ref prop setup to make it work.
